sdk/python/tests/test_BasicTransactionFactory.py

```diff
@@ -0,0 +1,115 @@
+			{
+				'general_parser': lambda x: self.general_parser_breadcrumbs.append(x) or x,
+				PublicKey: lambda x: self.public_key_parser_breadcrumbs.append(x) or x,
+			})
+
+		self.public_key_parser_breadcrumbs = []
+		self.general_parser_breadcrumbs = []
+		self.typenames_breadcrumbs = []
+
+	def create_by_name(self, typename):
+		self.typenames_breadcrumbs.append(typename)
+		return MockTransaction()
+
+	def create(self, descriptor):
+		# pass self as transaction factory
+		return self._create(descriptor, self)
+
+
+class BasicTransactionFactoryTest(unittest.TestCase):
+	def test_factory_requires_type(self):
+		# Arrange:
+		factory = MockTransactionFactory()
+
+		# Act + Assert:
+		with self.assertRaises(ValueError) as context:
+			factory.create({})
+
+		self.assertTrue('transaction descriptor does not have attribute type' in str(context.exception))
+
+	def test_factory_requires_type_and_signer_public_key(self):
+		# Arrange:
+		factory = MockTransactionFactory()
+
+		# Act + Assert:
+		with self.assertRaises(ValueError) as context:
+			factory.create({'type': 'one'})
+
+		self.assertTrue('transaction descriptor does not have attribute signer_public_key' in str(context.exception))
+
+	def test_factory_does_not_copy_type(self):
+		# Arrange:
+		factory = MockTransactionFactory()
+
+		# Act:
+		transaction = factory.create({'type': 'one', 'signer_public_key': 'two'})
+
+		# Assert:
+		self.assertEqual(['converted_one'], factory.typenames_breadcrumbs)
+		self.assertIsNone(transaction.type)
+		self.assertEqual(MockTransactionFactory.NETWORK, transaction.network)
+		self.assertEqual('converted_two', transaction.signer_public_key)
+
+	def test_factory_ignores_transaction_hints_for_both_type_and_signer_public_key(self):
+		# Arrange:
+		factory = MockTransactionFactory()
+
+		# Act:
+		transaction = factory.create({'type': 'one', 'signer_public_key': 'two'})
+
+		# Assert:
+		self.assertEqual(['converted_one'], factory.typenames_breadcrumbs)
+		self.assertEqual(['two'], factory.public_key_parser_breadcrumbs)
+		self.assertEqual([], factory.general_parser_breadcrumbs)
+
+		self.assertEqual('converted_two', transaction.signer_public_key)
+		self.assertEqual(None, transaction.breadcrumb_1)
+		self.assertEqual(None, transaction.breadcrumb_2)
+
+	def test_factory_applies_transaction_hints_for_other_fields(self):
+		# Arrange:
+		factory = MockTransactionFactory()
+		transaction = factory.create({'type': 'one', 'signer_public_key': 'two', 'breadcrumb_1': 'three', 'breadcrumb_2': 'four'})
+
+		# Assert:
+		self.assertEqual(['converted_one'], factory.typenames_breadcrumbs)
+		self.assertEqual(['two'], factory.public_key_parser_breadcrumbs)
+		# - parser has only been called for breadcrumb_1
+		self.assertEqual(['three'], factory.general_parser_breadcrumbs)
+
+		self.assertEqual('converted_two', transaction.signer_public_key)
+		self.assertEqual('converted_three', transaction.breadcrumb_1)
+		self.assertEqual('converted_four', transaction.breadcrumb_2)
```

Comment:
what is performing transition: four => converted_four

Reply:
type converter `lambda value: f'converted_{value}'`